### PR TITLE
Improve self-signed certificate generation:

### PR DESCRIPTION
**Pull request by nbougalis**
When starting, the code generates a new ephemeral private key and a corresponding certificate to go along with it. This process can take time and, while this is unlikely to matter for normal server operations, it can have a significant impact for unit testing and development. Profiling data suggests that ~20% of the time needed for a unit test run can be attributed to this.

This commit does several things:

1. It restructures the code so that a new self-signed certificate and its corresponding private key are only initialized once at startup; this has minimal impact on the operation of a regular server.
2. It provides new default DH parameters. This doesn't impact the security of the connection, but those who compile from scratch can generate new parameters if they so choose.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### CODE REVIEW

**Comment by nbougalis** [comment]
Generally, I think it might be a good idea to remove this automatic certificate generation from within the code. It's a convenient option to be sure, but it's also limited; for example, it will only generate RSA certificates. 

with options like LetsEncrypt for those whose servers have an associated domain, it is now trivially easy to get a "real" certificate from a well-known registrar, and it would be better if server operators did that instead.

Those who can't get a certificate from a CA could still generate a self-signed one. Information on how to do it on the web is plentiful but if necessary, we could include instructions on how to do it and it might even be something that could happen during initial package installation by invoking the right incantation.

File: src/ripple/basics/impl/make_SSLContext.cpp [hunk 1]
```diff
@@ -45,195 +82,88 @@ namespace detail {
...
-static x509_ptr
-x509_new()
-{
-    X509* x509 = X509_new();
+        if (x509 == nullptr)
+            LogicError("X509_new failed");
 
-    if (x509 == nullptr)
-        LogicError("X509_new failed");
+        X509_set_version(x509, NID_X509);
```
Comment (donovanhide) [inline src/ripple/basics/impl/make_SSLContext.cpp:131]: Still the wrong version:

https://github.com/ripple/rippled/issues/4007

Comment (nbougalis) [reply src/ripple/basics/impl/make_SSLContext.cpp:131]: Thanks for reminding me of this. I'll fix this.

**Comment by greg7mdp** [comment]
Maybe it is a naive question, because I didn't look at the code, but assuming most unit tests don't need the certificate, would it be possible to generate it the first time it is needed?

**Review by ximinez — APPROVED** [review]
I don't have a lot of SSL expertise, but these changes don't appear to be changing anything in that realm. The code is doing the same operations in the same order, only saving the self-signed certificate in a `static` context for easy reuse. 👍

**Comment by nbougalis** [comment]
> Maybe it is a naive question, because I didn't look at the code, but assuming most unit tests don't need the certificate, would it be possible to generate it the first time it is needed?

We would need to (a) restructure the startup sequence; and (b) make it possible to avoid instantiating the `Overlay` inside `ApplicationImp`. Both of these are good ideas and worth doing even if this is accepted & merged, but they are both much more complicated than this simple fix.

I folded a commit into this that addresses the concern by @donovanhide and fixes #4007. Apologies, I hadn't realized you had already reviewed, @ximinez, @greg7mdp.

**Comment by nbougalis** [comment]
Merged as 0ecfc7cb1a958b731e5f184876ea89ae2d4214ee

**Pull request closed without merging**